httpserver/handlers.go

```diff
@@ -58,35 +60,46 @@ func (server *Server) verify(ctx context.Context, w http.ResponseWriter, r *http
 		wg.Add(1)
 		go func(subject string) {
 			defer wg.Done()
-			// TODO: Enable caching:  Providers should add a caching mechanism to avoid extra calls to external data sources.
-			logrus.Infof("subject for request %v %v is %v", sanitizedMethod, sanitizedURL, subject)
-			returnItem := externaldata.Item{
-				Key: subject,
-			}
+			unlock := server.keyMutex.Lock(subject)
+			defer unlock()
 
-			verifyParameters := e.VerifyParameters{
-				Subject: subject,
-			}
-
-			result, err := server.GetExecutor().VerifySubject(ctx, verifyParameters)
-			if err != nil {
-				returnItem.Error = err.Error()
+			logrus.Infof("subject for request %v %v is %v", sanitizedMethod, sanitizedURL, subject)
+			var result types.VerifyResult
+			res := server.cache.get(subject)
+			if res != nil {
+				logrus.Debugf("cache hit for subject %v", subject)
+				result = *res
+			} else {
+				logrus.Debugf("cache miss for subject %v", subject)
+				returnItem := externaldata.Item{
+					Key: subject,
+				}
+
+				verifyParameters := executor.VerifyParameters{
+					Subject: subject,
+				}
+
+				if result, err = server.GetExecutor().VerifySubject(ctx, verifyParameters); err != nil {
+					returnItem.Error = err.Error()
+				}
+				server.cache.set(subject, &result)
+
+				if res, err := json.MarshalIndent(result, "", "  "); err == nil {
+					fmt.Println(string(res))
+				}
 			}
 
-			res, err := json.MarshalIndent(result, "", "  ")
-			if err == nil {
-				fmt.Println(string(res))
-			}
 			mu.Lock()
-			defer mu.Unlock()
```

Comment:
just curious (this may not be applicable at all): if we don't `defer` the unlock, would it be possible for the routine to get cancelled before the lock is released?

Reply:
that's a good question. I've changed it to defer again.